opentok-angular.js

```diff
@@ -311,3 +293,94 @@ ng.module('opentok', [])
       };
     }
   ]);
+
+  const captionSubscriberTracker = (OTSession, $rootScope) => {
+    const MAX_SUBS_ON_SCREEN = 4;
+    const CAPTIONS_TIMEOUT = 5 * 1000;
+
+    const captionBox = document.getElementById('caption-render-box');
+
+    let namesByConnectionId = {};
+
+    OTSession.session.on('signal:name', (event) => {
+      namesByConnectionId[event.from.connectionId] = event.data;
+    });
+    const getNameFromConnection = (connection) => {
+      let id = connection.creationTime.toString();
+      id = id.substring(id.length - 6, id.length - 1);
+      return `Guest${id}`;
+    };
+    const getName = (from) => {
+      if (!namesByConnectionId[from.connectionId]) {
+        namesByConnectionId[from.connectionId] = getNameFromConnection(from);
+      }
+      return namesByConnectionId[from.connectionId];
+    };
+
+    // we should remove the last element of the array if it exceeds the size
+    const captionsArray = $rootScope.captionsArray;
+    // the object should have the shape {caption, streamId, timeout, name}
+
+    // Actually adds the text to the box, maybe make the text smaller based on size?
+    const renderCaptionsArray = () => {
+      let captionString = '';
+      captionsArray.forEach((captionElm) => {
+        if (captionString === ''){
+          captionString = `${captionElm.name}: ${captionElm.captionText}`
+        } else {
+          captionString = `${captionString} \n ${captionElm.name}: ${captionElm.captionText}`
+        }
+      })
+      captionBox.innerText = captionString;
+    }
+
+    const alreadyHasStream = (streamId) => {
+      return (captionsArray.filter((elm) => elm.streamId === streamId)).length > 0;
+    }
+
+    // this function should just clear the timer and remove from the array, nothing else
+    const clearElementWithStreamId = (streamId) => {
+      const indexOfElement = captionsArray.findIndex((element) => {
+        return element.streamId === streamId
+      })
+
+      clearTimeout(captionsArray[indexOfElement].timeout)
+      captionsArray.splice(indexOfElement,1)
+    }
+
+    // We should remove the elemnt in question once we timeout and then rerender
+    const timeOutHandler = (streamId) => {
+      clearElementWithStreamId(streamId)
+      renderCaptionsArray();
+    }
+    const handleCaptionsEvent = (captionEvent,subscriber) => {
+      // if the streamId is already represented we push to the top and reset the timer
+      // otherwise we push this to the top and pop out and the last element and stop it's timer
+      // {captionText, streamId, timeout, name}
```

Comment:
What does this code do from the user's perspective?  It's not clear from your comment, IMO.